Adds folder_watcher component

robmarkcole opened the pull request:
## Description:
Replaces https://github.com/home-assistant/home-assistant/pull/12912 as becomes a standalone component. This component adds [Watchdog](https://github.com/gorakhargosh/watchdog) file system monitoring, with events published for changes to files within the filesystem.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io/pull/4885) with documentation (if applicable):** home-assistant/home-assistant.github.io#<4885>

## Example entry for `configuration.yaml` (if applicable):
```yaml
folder_watcher:
  - folder: /Users/robincole/.homeassistant/images
    patterns:
      - '*.jpg'
  - folder: /Users/robincole/.homeassistant/www
    patterns:
      - '*.jpg'
```

## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

robmarkcole committed:
Create watchdog_file_watcher.py

OttoWinter commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+DOMAIN = "watchdog_file_watcher"
+EVENT_TYPE = "event_type"
+SRC_PATH = "src_path"
+PATTERNS = ["*.txt", "*.py", "*.md", "*.jpg", "*.png"]
```
I don't think we should restrict file patterns like this, if at all.

dgomes replied:
Make it configurable

robmarkcole replied:
Agreed

OttoWinter commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+        _LOGGER.error("folder %s is not valid or allowed", path)
+    else:
+        Watcher(path, hass)
+    return True
```
I think we should `return False` in the `_LOGGER.error("folder %s is not valid or allowed", path)` block.

robmarkcole replied:
ok

OttoWinter commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+def async_setup(hass, config):
+    """Set up the folder watcher."""
+    conf = config.get(DOMAIN)
+    path = conf.get(CONF_PATH)
```
Use index syntax for required keys:

```python
path = conf[CONF_PATH]
```

robmarkcole replied:
ok

OttoWinter commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+        self.hass.bus.fire(
+            DOMAIN, {
+                EVENT_TYPE: event.event_type,
+                SRC_PATH: event.src_path})
```
Just a question, but should we use absolute paths for events or paths relative to the directory we specified.

robmarkcole replied:
Well my use case is having an automation run on the file identified by the path, so if we use relative paths is would then require the automation to append the rest of the path which is a hassle. I propose just splitting out the file_name and firing that too?

OttoWinter commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+    if not hass.config.is_allowed_path(path):
+        _LOGGER.error("folder %s is not valid or allowed", path)
+    else:
+        Watcher(path, hass)
```
I'm not entirely sure how entities are added for components, but when does this entity get added to hass?

Also, why is this a component and not a `sensor` platform?

robmarkcole replied:
Im not sure about this too, but it appears to work..

OttoWinter commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+        """Process the Watchdog event."""
+        self.hass.bus.fire(
+            DOMAIN, {
+                EVENT_TYPE: event.event_type,
```
I think you need to specify the event_type as a positional arg/kwarg. [`fire` signature](https://github.com/home-assistant/home-assistant/blob/e7b84432f94ffdd0d1e05550e8f2d506f23ae526/homeassistant/core.py#L380):

```python
def fire(self, event_type: str, event_data=None, origin=EventOrigin.local):
```

Also, I don't think we should be directly using the `event_type`s from watchdog, we should at least translate them to `EVENT_*` constants even if they end up being the same string. I mean it probably won't happen, but what if watchdog changes their event types.

balloob replied:
Well Otto, that's where Rob is one step ahead of you 😉 He is actually firing `watchdog_file_watcher` events. The subtype is the watchdog type.

I am not sure if we would want to have 1 event with sub event types or have each watchdog event type be it's own event in Home Assistant. I kinda like how it is now.

robmarkcole replied:
I want automations to trigger on particular event types and not make this too complicated, I also like how it is now

OttoWinter commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+    return True
+
+
+class Watcher(Entity):
```
I think you can move the event handler into the entity:

```python
class Watcher(Entity, PatternMatchingEventHandler):
```

robmarkcole replied:
I will have a go

OttoWinter commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+            MyHandler(hass),
+            path,
+            recursive=True)
+        self._observer.start()
```
I think we should only start observing once we're `added_to_hass`. Doing work or I/O in the constructor is a really bad idea 😉

balloob replied:
There is no `added_to_hass` as it's not an entity. Instead, we should listen to the HOMEASSISTANT_START event and start then.

robmarkcole replied:
ok

balloob commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+from homeassistant.helpers.entity import Entity
+import homeassistant.helpers.config_validation as cv
+
+#DEPENDENCIES = ['watchdog']
```
Remove this.

robmarkcole replied:
ok

balloob commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+"""
+Component for monitoring activity on a folder.
```
The name of this component is weird.

 - it's watching a folder
 - but it's called watchdog_file_watcher

Watchdog is the requirement that is used. Probably just `folder_watcher` would be an ok name ?

balloob replied:
Also don't forget to specify a `REQUIREMENTS` constant with watchdog in it

robmarkcole replied:
Yes I'm happy with `folder_watcher`

balloob commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+                EVENT_TYPE: event.event_type,
+                SRC_PATH: event.src_path})
+
+    def on_modified(self, event):
```
What happened to the `on_any_event` handler ?

balloob commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
```
This should not be async. You're doing a shit ton of I/O which is not allowed inside the event loop.

robmarkcole replied:
ok

balloob commented on homeassistant/components/watchdog_file_watcher.py:
```diff
@@ -0,0 +1,75 @@
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_PATH): cv.isdir})
```
This only allows 1 path to be specified. I think that we should take in a list of paths. `vol.All(cv.ensure_list, [cv.isdir])`

robmarkcole replied:
Agreed. I propose to take a list of paths, with recursive and patterns as optional on each path.

robmarkcole committed:
Rename watchdog_file_watcher.py to folder_watcher.py

robmarkcole commented:
Thanks for your feedback @dgomes and @OttoWinter I have time to work on this at the weekend. 
Cheers

robmarkcole committed:
Address a number of issues

robmarkcole commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,85 @@
+            MyHandler(hass),
+            path,
+            recursive=DEFAULT_IS_RECURSIVE)
+        self._observer.start()
```
How to listen for HOMEASSISTANT_START - I have to apply a decorator to `self._observer.start()`?

OttoWinter replied:
I didn't really understand @balloob's comment from before, since from what I know this *is* an entity. Here's how I would have done it:

```python
class FolderWatcher(Entity, FileSystemEventHandler):
    def __init__(self, path):
        self._observer = Observer()
        self._path = path

    async def async_added_to_hass(self):
        self.hass.async_add_job(self.start_watching)

    async def async_will_remove_from_hass(self):
        self.hass.async_add_job(self.stop_watching)

    def start_watching(self):
        self._observer.schedule(self, self._path, recursive=True)
        self._observer.start()

    def stop_watching(self):
        self._observer.stop()

    def process(self, event):
        on_any_event(self.hass, event)

    def on_modified(self, event):
        self.process(event)

    def on_moved(self, event):
        self.process(event)

    def on_created(self, event):
        self.process(event)

    def on_deleted(self, event):
        self.process(event)
```

OttoWinter replied:
(My GitHub client is not working correctly, repost)

I didn't really understand @balloob's comment from before, since from what I know this *is* an entity. Here's how I would have done it:

```python
class FolderWatcher(Entity, FileSystemEventHandler):
    def __init__(self, path):
        self._observer = Observer()
        self._path = path

    async def async_added_to_hass(self):
        self.hass.async_add_job(self.start_watching)

    async def async_will_remove_from_hass(self):
        self.hass.async_add_job(self.stop_watching)

    def start_watching(self):
        self._observer.schedule(self, self._path, recursive=True)
        self._observer.start()

    def stop_watching(self):
        self._observer.stop()

    def process(self, event):
        on_any_event(self.hass, event)

    def on_modified(self, event):
        self.process(event)

    def on_moved(self, event):
        self.process(event)

    def on_created(self, event):
        self.process(event)

    def on_deleted(self, event):
        self.process(event)
```

robmarkcole replied:
@OttoWinter tried that but async_added_to_hass() is never called

robmarkcole replied:
I also receive another error using `FolderWatcher(Entity, FileSystemEventHandler):` that I don't understand:
```
2018-03-08 06:36:30 ERROR (MainThread) [homeassistant.setup] Error during setup of component folder_watcher
Traceback (most recent call last):
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/setup.py", line 145, in _async_setup_component
    component.setup, hass, processed_config)
  File "/anaconda3/lib/python3.6/asyncio/futures.py", line 332, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/anaconda3/lib/python3.6/asyncio/tasks.py", line 250, in _wakeup
    future.result()
  File "/anaconda3/lib/python3.6/asyncio/futures.py", line 245, in result
    raise self._exception
  File "/anaconda3/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/robincole/.homeassistant/custom_components/folder_watcher.py", line 38, in setup
    FolderWatcher(path)
  File "/Users/robincole/.homeassistant/custom_components/folder_watcher.py", line 57, in __init__
    self._observer.schedule(self, self._path, recursive=True)
  File "/Users/robincole/Documents/Github/home-assistant/lib/python3.6/site-packages/watchdog/observers/fsevents.py", line 172, in schedule
    return BaseObserver.schedule(self, event_handler, path, recursive)
  File "/Users/robincole/Documents/Github/home-assistant/lib/python3.6/site-packages/watchdog/observers/api.py", line 284, in schedule
    self._add_handler_for_watch(event_handler, watch)
  File "/Users/robincole/Documents/Github/home-assistant/lib/python3.6/site-packages/watchdog/observers/api.py", line 243, in _add_handler_for_watch
    self._handlers[watch].add(event_handler)
TypeError: unhashable type: 'FolderWatcher'
```

OttoWinter replied:
Are you using the `setup_`**`platform()`** code? I'm pretty sure `async_added_to_hass` would be called for entities of a platform component and quite a few integrations depend on it.

```python
def setup_platform(hass, config, add_devices, discovery_info=None):
    # ...
    add_devices([FolderWatcher(...)])
```

That *should* call the `async_added_to_hass` callback.

OttoWinter replied:
Oh the unhashable type is an issue. If I understand correctly `watchdog` tries to put the entity in a `set`. And to be added to a set an object must be hashable (implementing the hash function). Entities however are apparently unhashable, so using `class FolderWatcher(Entity, FileSystemEventHandler):` won't work.

robmarkcole replied:
@OttoWinter I'm using setup(), also I am confused now why you recommend setup_platform() as this is a component, not a sensor, perhaps I'm missing something?

OttoWinter replied:
Ohhhh! Sorry I thought this would be a replacement for the folder sensor, just with the added benefit of firing events. However, you correctly said "Replaces #12912 as becomes a standalone component." at the top of your PR.

balloob replied:
If you want to learn how to use the start event, check the code… `git grep EVENT_HOMEASSISTANT_START`

robmarkcole commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,85 @@
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_FOLDERS):
+            vol.All(cv.ensure_list, [cv.isdir]),
+        })
```
Could add config for file types to watch/not_watch - should this be applied to all folders or folder by folder? I'm not a fan of overly complex config

OttoWinter replied:
I think if this should be added, then as a [glob-style](https://docs.python.org/3/library/glob.html) filter, though I don't think the it's really something we should do anyway.

robmarkcole replied:
@OttoWinter what is your thinking behind not adding?

robmarkcole commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,85 @@
+
+
+def on_any_event(hass, event):
+    if not event.is_directory:
```
Added otherwise you get a `folder modified` event every time a file is changed, which is overkill in my opinion but perhaps people want this?

robmarkcole commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,85 @@
+        self._observer.schedule(
+            MyHandler(hass),
+            path,
+            recursive=DEFAULT_IS_RECURSIVE)
```
Make IS_RECURSIVE configurable?

OttoWinter replied:
I personally don't think so, but it could be also done using the `glob` approach described before.

robmarkcole commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,85 @@
+                    'folder': event_folder})
+
+
+class MyHandler(FileSystemEventHandler):
```
Probably EventHandler is a better name - but don't want to confuse with Home-assistant events..? Alternatively as per @OttoWinter previous comments I could merge this class with the `Watcher` class, although the current implementation makes comparison with the Watchdog examples very clear

robmarkcole replied:
Changed to `WatcherHandler`

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,85 @@
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
```
Now that you've converted this into a platform, you also need to change this to a `setup_platform` function. See this for an example:

https://github.com/home-assistant/home-assistant/blob/c462292e4d3017401a7def24ca8051b34ad605f1/homeassistant/components/sensor/folder.py#L45-L53

robmarkcole replied:
Oh yes, I wrote that :)

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,85 @@
+_LOGGER = logging.getLogger(__name__)
+
+CONF_FOLDERS = 'folders'
+DOMAIN = "folder_watcher"
```
Platforms don't have domains.

robmarkcole replied:
Is component

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,85 @@
+from homeassistant.helpers.entity import Entity
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['watchdog']
```
I think we're required to pin the requirements versions. For example

```python
REQUIREMENTS = ['watchdog==0.8.3']
```

robmarkcole replied:
ok

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,85 @@
+import os
+import logging
+import voluptuous as vol
+from watchdog.observers import Observer
```
Then also move these imports into local imports. For example https://github.com/home-assistant/home-assistant/blob/c462292e4d3017401a7def24ca8051b34ad605f1/homeassistant/components/media_player/cast.py#L355

robmarkcole replied:
@OttoWinter If I do this:
![image](https://user-images.githubusercontent.com/11855322/37137377-5d92e8d2-229d-11e8-9623-e88c40a1787e.png)

I get the error: 
```
  File "/Users/robincole/.homeassistant/custom_components/folder_watcher.py", line 63, in <module>
    class WatcherHandler(FileSystemEventHandler):
NameError: name 'FileSystemEventHandler' is not defined
```

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,85 @@
+EVENT_TYPE = "event_type"
+IS_DIRECTORY = "is_directory"
+
+CONFIG_SCHEMA = vol.Schema({
```
As a platform, you now need to specify a `PLATFORM_SCHEMA` instead:

https://github.com/home-assistant/home-assistant/blob/c462292e4d3017401a7def24ca8051b34ad605f1/homeassistant/components/sensor/folder.py#L26-L29

robmarkcole committed:
Adds filter

robmarkcole commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,93 @@
+import os
+import logging
+import voluptuous as vol
+from watchdog.events import FileSystemEventHandler
```
I'm not sure where to import

robmarkcole commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,93 @@
+    if not event.is_directory:
+        file_name = os.path.split(event.src_path)[1]
+        if not file_name.startswith('.'):  # Avoid hidden files.
+            if fnmatch.fnmatch(file_name, file_filter):  # Apply filter.
```
In my ignorance of fnmatch I am unsure how to apply multiple filters, e.g. *.txt *.jpg

robmarkcole commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,93 @@
+            recursive=True)
+        self._observer.start()
+
+    def stop_watching(self):
```
This is currently never called

robmarkcole commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,93 @@
+    """On Watcher event, fire HA event."""
+    if not event.is_directory:
+        file_name = os.path.split(event.src_path)[1]
+        if not file_name.startswith('.'):  # Avoid hidden files.
```
I'm now filtering out hidden files, which I presume people prefer?

robmarkcole commented:
@balloob @dgomes  @OttoWinter I've addressed several of your comments but still unclear on startup or imports. 

I've implemented file filter in identical fashion to the folder sensor but it's a bit clunky, allows only a single file filter, and applies the same filter to every folder watched. Still it's better than nothing?

Finally I still don't have an answer [to this question](https://community.home-assistant.io/t/usage-of-trigger-event-data-in-notification/45560), which sort of defeats the point of the whole component?
cheers

OttoWinter commented:
Sorry about those previous comments, a few of them do not apply since this a component and not a platform. Here's a few thoughts I have on this:

## Use cases

I think we should think about use cases for this component for our other decisions.

* Use case #.1: Reloading/Restarting Home Assistant automatically on configuration file changes.
  * Would need some sort of **filtering files**, since we shouldn't restart Home Assistant on `entity_registry.yaml` updates.
* Use case #.2: File system events from *other applications* (for example backups).
  * Will probably never work with hass.io/docker, since the file systems are intentionally separated for those.
* I'm sure there are other uses cases for this, feel free to comment them.

## File filtering

You implemented some file filtering options, I understand the need for them, but we should think about *how* they should be done.

* Option #.1: Don't filter files. This was my initial idea to keep the component as simple as possible, but use cases #.1 & #.2 heavily depend on filtering files.
  * I think if we **do not** allow manual filtering, we should definitely **not** filter *some* files by default, for example hidden files (`.*`). At some point the component just seems to restrictive.
* Option #.2: File filtering **inside** FolderWatcher. Things to consider:
  * The configuration would be cleaner. Having the filters right next to the `folder_watcher` (or whatever it will be called) is nice.
  * What if I want a folder watcher `A` for `/config` with filtering out `/config/foo.txt`, but another folder watcher also for `/config`, but without any filters. Because we publish to the event bus *globally*, we can't distinguish between the events of the two.
 * Option #.3: File filtering in **automations**.
   * Filters would sit far away from the `folder_watcher:` config. Not very nice.
   * Not sure how this *could* be done at all. Maybe with a new `condition:` type, but adding yet 
another condition type is not too nice.
   * Allows automations to watch for very specific file events.

File filters we could use:
 * [globbing](https://docs.python.org/3/library/glob.html): `**/*.py`
   * Can't be used very well with multiple filters. For example filtering out `entity_registry.yaml`, but not `configuration.yaml`.
   * watchdog has something like this already: [PatternMatchingEventHandler](http://pythonhosted.org/watchdog/api.html#watchdog.events.PatternMatchingEventHandler)
 * [.gitignore style](https://git-scm.com/docs/gitignore)
   * Well known and easy to use. Still allows for complicated filters.
 * regex
   * Quite well known
   * Can be difficult to use
   * Has "too many" options (also not sure security-wise)
   * Usually not used for paths.

## Platform vs. Component

At first I thought this was going to be a rewrite of the folder sensor. Honestly, I think it would kind of make for sense there. Maybe there already was a discussion on this that I missed/can't remember. Some thoughts:

* With a `folder` sensor **and** `folder` component, I think we would only confuse users. They seem quite similar functionality-wise to me, only with this PR having the events.
* On the other side, I don't really expect `sensor`s to fire events...

## `watchdog` import issues

We can't do global imports because then every user would need to download `watchdog` even if they're not using the folder watcher. Your case is however a bit tricky, since you need to create a class that extends an imported abstract class.

This could be a way to fix it; it's not very nice though and there might very well be better options.

```python
def create_event_handler(file_filter, hass):
    from watchdog.events import FileSystemEventHandler

    class FolderWatcherEventHandler(FileSystemEventHandler):
        def __init__(self, file_filter, hass):
            super().__init__(FolderWatcherEventHandler, self).__init__()
            self._file_filter = file_filter
            self.hass = hass

        def process(self, event):
            on_any_event(self.hass, self._file_filter, event)

        def on_modified(self, event):
            self.process(event)

        def on_moved(self, event):
            self.process(event)

        def on_created(self, event):
            self.process(event)

        def on_deleted(self, event):
            self.process(event)

    return FolderWatcherEventHandler(file_filter, hass)
```

robmarkcole commented:
@OttoWinter thanks for your detailed comments. 

## Use case
Mine is to monitor a folder where a motion activate camera is saving time-stamped files. I will use an  automation to watch for new images then send them image processing. Again if anyone can answer my [question](https://community.home-assistant.io/t/usage-of-trigger-event-data-in-notification/45560) on how to that I would appreciate it :) The same  principle applies to services that save .txt or other files. 

## File filtering
I agree this is necessary. I was previously using the PatternMatchingEventHandler class which uses glob I believe. I will use this class as it's making best use of Watchdog and consistent with the folder sensor. 
I propose making config:
```
folder_watcher:
  filters: 
    -  '*.jpg'
    -  '*.txt'
  folders:
    - /.homeassistant/images
    - /.homeassistant/www
```
Re the hidden files, perhaps its just 
## Platform vs. Component
There was indeed an earlier conversation about this. A sensor might miss files being added if two were added at the same time, therefore events are necessary. 
The folder sensor has a slightly different use case - it is for when you want to monitor the number of files in a folder, and the size of the folder. 

Re imports, I'l try your suggestion

robmarkcole committed:
Adds pattern matching

robmarkcole committed:
Adds create_event_handler()

balloob commented on homeassistant/components/folder_watcher.py:
```diff
@@ -1,66 +1,68 @@
 
 
-def on_any_event(hass, file_filter, event):
+def on_any_event(hass, event):
```
Why extract this out into a method ?

robmarkcole replied:
Replaced with process()

robmarkcole commented:
# ASSISTANTCE REQUEST
As per [this post](https://community.home-assistant.io/t/usage-of-trigger-event-data-in-notification/45560), I'm currently unable to trigger an automation using this component. I always get the error:
```
Traceback (most recent call last):
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/core.py", line 1006, in _event_to_service_call
    await service_handler.func(service_call)
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/components/notify/__init__.py", line 134, in async_notify_message
    kwargs[ATTR_MESSAGE] = message.async_render()
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/helpers/template.py", line 133, in async_render
    raise TemplateError(err)
homeassistant.exceptions.TemplateError: UndefinedError: 'trigger' is undefined
```
If this component needs changing to resolve this issue I hope someone can advise. 
There are a couple of open issues which may be related:
1. https://github.com/home-assistant/home-assistant/issues/12755
2. https://github.com/home-assistant/home-assistant/issues/9618

robmarkcole committed:
Update folder_watcher.py

robmarkcole committed:
Adds run_setup()

robmarkcole committed:
Remove stop_watching()

balloob commented:
There is this great app called Google. Type in "Home Assistant Automation Template" (without the quotes)

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,96 @@
+            create_event_handler(patterns, hass),
+            path,
+            recursive=True)
+        self._observer.start()
```
If I understand correctly, this will create a new thread (`threading.Thread`). We need to stop that thread when shutting down Home Assistant, because otherwise we would significantly slow down shut down.

For example, the cast integration did the same for a long time. It created a new thread for each discovered cast and never closed them on Home Assistant shutdown. This made Home Assistant hang for about 10s before shutting down after calling `homeassistant.stop`.

So you should additionally listen to `EVENT_HOMEASSISTANT_STOP` and clean up the threads there with something like

```
observer.stop()
# maybe also call "observer.join()", though that may not 
# actually be necessary because we don't care about the result.
```

OttoWinter replied:
Oh, it seems like you had this in another commit, but then reverted it... Was there a reason/a discussion for that I missed? Is `watchdog` handling this for us already?

robmarkcole replied:
OK this appears to work:
```python
class Watcher(Entity):
    """Class for starting Watchdog."""
    def __init__(self, path, patterns, hass):
        from watchdog.observers import Observer
        self._observer = Observer()
        self._observer.schedule(
            create_event_handler(patterns, hass),
            path,
            recursive=True)
        self._observer.start()
        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self.shutdown)

    def shutdown(self, event):
        """Shutdown the watcher."""
        self._observer.stop()
```

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,96 @@
+            """On Watcher event, fire HA event."""
+            if not event.is_directory:
+                file_name = os.path.split(event.src_path)[1]
+                folder_name = os.path.split(event.src_path)[0]
```
You could just write this (tiny bit more efficient and easier to read 🤓)

```python
folder_path, file_name = os.path.split(event.src_path)
```

robmarkcole replied:
nice

robmarkcole commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,96 @@
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_FOLDERS):
+            vol.All(cv.ensure_list, [cv.isdir]),
+        vol.Optional(CONF_FILTERS, default=[DEFAULT_FILTER]):
```
Re config yes I will look at that.

I agree that `filters` is ambiguous. [PatternMatchingEventHandler](https://pythonhosted.org/watchdog/api.html?highlight=patternmatchingeventhandler#watchdog.events.PatternMatchingEventHandler) uses `patterns` and `ignore_patterns` so this should do the same.

robmarkcole committed:
Adds shutdown()

robmarkcole committed:
Update config to allow patterns on each folder

robmarkcole committed:
Update to patterns from filters

robmarkcole commented:
@OttoWinter the closest I could find to your suggested config is the [multiple hue bridge](https://home-assistant.io/components/hue/#multiple-hue-bridges).

robmarkcole commented:
I'm thinking of changing the naming in the config to something like:
```
folder_watcher:
  folders:
    - path: /config
      patterns:
        - '*.yaml'
    - path: /www
      patterns:
         - '*.jpg'
```
Any suggestions @OttoWinter ?

robmarkcole committed:
Adds watchdog

robmarkcole committed:
Fix indents on schema

robmarkcole committed:
Update folder_watcher.py

robmarkcole committed:
Create test_file_watcher.py

robmarkcole committed:
Fix lints

robmarkcole committed:
Add test_invalid_path()

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,118 @@
+                _LOGGER.error("folder %s is not valid or allowed", path)
+                continue
+            else:
+                Watcher(path, patterns, hass)
```
Unnecessary `else` since you `continue` above.

```python
if not hass.config.is_allowed_path(path):
    _LOGGER.error("folder %s is not valid or allowed", path)
    continue
Watcher(path, patterns, hass)
```

robmarkcole replied:
Hmm the component doesn't setup at all if I do that, will drop the `continue`

robmarkcole replied:
ok

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,118 @@
+            else:
+                Watcher(path, patterns, hass)
+
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, run_setup)
```
I'm not sure if this is necessarily the correct way. Now we only add the sensors when Home Assistant is started, but I think we should only *register the observer* when Home Assistant is started. I could very well be wrong about this, but I think we should initialize entities at `setup` time, and not at `EVENT_HOMEASSISTANT_START` time.

```python
# in setup()
conf = config[DOMAIN]
watchers = conf[CONF_WATCHERS]
for watcher in watchers:
    path = watcher[CONF_FOLDER]
    patterns = watcher[CONF_PATTERNS]
    if not hass.config.is_allowed_path(path):
        _LOGGER.error("folder %s is not valid or allowed", path)
        continue
    Watcher(path, patterns, hass)
```

```python
# in Watcher#__init__
self._observer = Observer()
hass.bus.listen_once(EVENT_HOMEASSISTANT_START, self.startup)
hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self.shutdown)
```

```python
# in Watcher#startup
self._observer.schedule(
    create_event_handler(patterns, hass),
    path,
    recursive=True)
self._observer.start()
```

robmarkcole replied:
@OttoWinter I've tried that out but `self._observer.start()` is never called - presumably because `Watcher#__init__` is called by `EVENT_HOMEASSISTANT_START` and `self.startup` is listening for another `EVENT_HOMEASSISTANT_START` which will never come

OttoWinter commented on tests/components/test_file_watcher.py:
```diff
@@ -0,0 +1,59 @@
+def create_file(path):
+    """Create a test file."""
+    with open(path, 'w') as test_file:
+        test_file.write("test")
```
I don't really like tests that do actual I/O, especially when it's not too important to do that. This could end up being one of those pesky flakey tests. From what I see, `watchdog` itself is well tested and so there's no need for us to test `watchdog` in  Home Assistant too. We should only test whether the callbacks we *receive* from `watchdog` are properly handled by *our* code and that our setup methods call `watchdog` methods correctly. 

Maybe have a look at the mocking framework for how to do this best: https://docs.python.org/3/library/unittest.mock.html

robmarkcole replied:
I copied this from the [folder sensor test](https://github.com/home-assistant/home-assistant/blob/dev/tests/components/sensor/test_folder.py), but agree this is covered by `watchdog` testing and could be omitted.

OttoWinter commented on tests/components/test_file_watcher.py:
```diff
@@ -0,0 +1,59 @@
+
+
+class TestFolderWatcher(unittest.TestCase):
+    """Test the filesize sensor."""
```
Stale docstring

robmarkcole replied:
ok

OttoWinter reviewed (commented):
Not sure about the config format, why would we want an additional level of indentation when there's only one thing we're doing anyway?

```python
folder_watcher:
  folders: # <- why this?
    - path: whatever
      patterns:
        - hi
        - there*
```

robmarkcole committed:
Adds folder_watcher

robmarkcole commented:
@OttoWinter RE your comment about the additional level of indentation in the config, the method I'm using is the only way I can get it to work, following the example of the hue bridge. I  searched many components to see if any where using the config format you suggest but unfortunately couldn't find any.

robmarkcole committed:
Update test_file_watcher.py

robmarkcole committed:
Update folder_watcher.py

robmarkcole committed:
Simplify config

robmarkcole committed:
Adapt for new config

robmarkcole committed:
Run observer.schedule() on EVENT_HOMEASSISTANT_START

robmarkcole commented:
@OttoWinter I think just tests to fix now?

OttoWinter commented:
Ok, here's how I would have done it: https://gist.github.com/OttoWinter/24ec8b06d7b3805d2ff96b1c6ddeefda

A couple of things going on here:
* `CONFIG_SCHEMA`: We have to use `cv.ensure_list`
* `Watcher` class is never added to Home Assistant as an Entity and doesn't expose any state. So I don't understand why it would be a subclass of `Entity`.
* `PatternMatchingEventHandler` defaults to case-insensitive mode. Do we want that?
* `FileSystemEventHandler` has a method `on_any_event`. Should we use that or our own maybe safer way?
* `Observer#schedule()` doesn't do any I/O, so it's safe to call from the constructor. (Got that wrong before, sorry)

robmarkcole committed:
Amend Watcher removing entity and tidying startup

robmarkcole commented:
@OttoWinter it appears that `hass.bus.listen_once` passes `event` to startup and I get an error unless I do the following:
```
        hass.bus.listen_once(EVENT_HOMEASSISTANT_START, self.startup)

    def startup(self, event):
        """Start the watcher."""
        self._observer.start()
```

robmarkcole committed:
Tidy config

robmarkcole committed:
Rename process to on_any_event for consistency

robmarkcole committed:
Rename on_any_event back to process

Using `on_any_event` resulted in 2 events being fired

robmarkcole commented:
Thanks @OttoWinter renamed back to `process`. Any more comments on functionality or should I progress to tests?
Btw this works on Hassio too :-)

robmarkcole commented:
@OttoWinter having used this component for real I propose just firing the src_path rather than the file_name and folder as src_path is typically what you want for automations. People can then use templates to extract file_name if they wish, but really I can't see much use for the file_name. Any thoughts? Any other info we might want to fire in the event?

OttoWinter commented:
I mean the advantage of separate folder and file is that is really easy to stitch them together again, e.g.

```
{{ trigger.event.data.folder }}/{{ trigger.event.data.file }} 
```

The other way around, so converting from path to folder+file is more difficult (albeit not too much):

```
Folder: {{ trigger.event.data.path.split('/')[:-1] | join('/') }}
File: {{ trigger.event.data.path.split('/')[-1] }}
```

Just a thought, but we could maybe send both. It wouldn't be very nice though.

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,111 @@
+        patterns = watcher[CONF_PATTERNS]
+        if not hass.config.is_allowed_path(path):
+            _LOGGER.error("folder %s is not valid or allowed", path)
+            continue
```
Hmmm, should we maybe consider returning `False` (so creating a persistent notification saying we weren't able to set up)? I mean few users actually look through the logs and so maybe some would not notice their folder watcher not being set up. 

With this integration it would be especially hard to debug the issue as a user (if you're not looking at the logs) because you don't have any immediate visual feedback that setup worked or not. For example if you create a folder watcher for an unallowed path, you would just not receive the events. That might confuse some people.

Is there any use case for breaking kind of silently anyway?

robmarkcole replied:
Adopting the `False `

OttoWinter reviewed (commented):
Yeah I think the code for this component is pretty much done 👍

Now only the tests remain... (always fun writing those 😥)

robmarkcole committed:
Update folder_watcher.py

robmarkcole commented:
@OttoWinter added the `path` to the event data, I think its equally likely you would want path, file or folder so let's have them all and avoid nasty templates that will definitely lead to several forum threads..
Onto the unit tests..!

OttoWinter commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,109 @@
+        patterns = watcher[CONF_PATTERNS]
+        if not hass.config.is_allowed_path(path):
+            _LOGGER.error("folder %s is not valid or allowed", path)
+            False
```
`return False`? 👀 (this ain't ruby 🙈)

robmarkcole replied:
return True ;-)

robmarkcole commented:
@OttoWinter I am having a headache with the tests. Ive added watchdog to `script/gen_requirements_all.py`, now on running `tox -r` I am hitting the error: `ERROR: could not load /Users/robincole/Documents/Github/home-assistant/tests/conftest.py`. Any advice? Full traceback below:
```
py36 runtests: commands[0] | py.test --timeout=9 --duration=10 --cov --cov-report= tests/components/test_folder_watcher.py
Traceback (most recent call last):
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/_pytest/config.py", line 328, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/Users/robincole/Documents/Github/home-assistant/tests/components/test_folder_watcher.py')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/_pytest/config.py", line 328, in _getconftestmodules
    return self._path2confmods[path]
KeyError: local('/Users/robincole/Documents/Github/home-assistant/tests/components')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/_pytest/config.py", line 359, in _importconftest
    return self._conftestpath2mod[conftestpath]
KeyError: local('/Users/robincole/Documents/Github/home-assistant/tests/conftest.py')

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/_pytest/config.py", line 365, in _importconftest
    mod = conftestpath.pyimport()
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/py/_path/local.py", line 668, in pyimport
    __import__(modname)
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/_pytest/assertion/rewrite.py", line 212, in load_module
    py.builtin.exec_(co, mod.__dict__)
  File "/Users/robincole/Documents/Github/home-assistant/tests/conftest.py", line 14, in <module>
    from tests.common import async_test_home_assistant, INSTANCES, \
  File "/Users/robincole/Documents/Github/home-assistant/tests/common.py", line 13, in <module>
    from homeassistant import core as ha, loader, config_entries
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/config_entries.py", line 120, in <module>
    from .setup import async_setup_component, async_process_deps_reqs
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/setup.py", line 9, in <module>
    from homeassistant import requirements, core, loader, config as conf_util
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/requirements.py", line 7, in <module>
    import homeassistant.util.package as pkg_util
  File "/Users/robincole/Documents/Github/home-assistant/homeassistant/util/package.py", line 10, in <module>
    from pip.locations import running_under_virtualenv
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/pip/__init__.py", line 45, in <module>
    from pip.vcs import git, mercurial, subversion, bazaar  # noqa
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/pip/vcs/mercurial.py", line 9, in <module>
    from pip.download import path_to_url
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/pip/download.py", line 40, in <module>
    from pip._vendor import requests, six
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/pip/_vendor/requests/__init__.py", line 98, in <module>
    from . import packages
  File "/Users/robincole/Documents/Github/home-assistant/.tox/py36/lib/python3.6/site-packages/pip/_vendor/requests/packages.py", line 12, in <module>
    sys.modules['pip._vendor.requests.packages.' + mod] = sys.modules["pip._vendor." + mod]
KeyError: 'pip._vendor.urllib3.contrib'
ERROR: could not load /Users/robincole/Documents/Github/home-assistant/tests/conftest.py

ERROR: InvocationError: '/Users/robincole/Documents/Github/home-assistant/.tox/py36/bin/py.test --timeout=9 --duration=10 --cov --cov-report= tests/components/test_folder_watcher.py'
____________________________________________________________________________________________ summary ____________________________________________________________________________________________
ERROR:   py36: commands failed
```

robmarkcole committed:
Fix return False on setup

robmarkcole committed:
Update test_file_watcher.py

robmarkcole commented:
Travis is failing on import error although watchdog has been added to `requirements_all.txt`, any ideas?

robmarkcole commented:
## Tests
@OttoWinter as per your earlier suggestion the appropriate test is to mock the watchdog `event` object. TBH I have no idea how to do this correctly. Should I call `create_event_handler()` and pass a mock `event` object the `process()` method? Or should I somehow mock the `event` object firing on the system bus? Any advice gratefully received, also @balloob @dgomes

dgomes commented on homeassistant/components/folder_watcher.py:
```diff
@@ -0,0 +1,109 @@
+    """Class for starting Watchdog."""
+
+    def __init__(self, path, patterns, hass):
+        """Initialise the Watchdog oberver."""
```
observer

robmarkcole replied:
ok

robmarkcole committed:
Update folder_watcher.py

OttoWinter commented:
@robmarkcole I mean I really don't care too strongly about actually creating a file and testing that way. It just feels like `watchdog` is thoroughly tested already and we should only test our code in the tests, not watchdogs code.

About the error: I believe it's because the `watchdog` is not installed in the tox test environment, AFAIK `tox` only installs the requirements in `requirements_test_all.txt` (not `requirements_all.txt`) for tests. But ideally, again, you wouldn't even need to import watchdog, you can just "mock" watchdog and check the calls. Take a look at how the ~~chromecast~~ cast integration tests do it:

https://github.com/home-assistant/home-assistant/blob/947218d51c94566feeb32c6f549493fa07683252/tests/components/media_player/test_cast.py#L16-L22

robmarkcole committed:
Adds watchdog

robmarkcole commented:
@OttoWinter re the error, I've added to `requirements_test_all.txt` and run `pip3 install -r requirements_test_all.txt` and `tox -r` but still the error persists.

robmarkcole committed:
Undo adding watchdog

robmarkcole committed:
Update test_file_watcher.py

balloob commented:
For your tests, we should never do I/O in tests. We should mock all I/O and only test the interactions that get triggered by I/O happening.

robmarkcole committed:
Update test_file_watcher.py

robmarkcole commented on tests/components/test_file_watcher.py:
```diff
@@ -0,0 +1,52 @@
+        self.assertFalse(
+            setup_component(self.hass, DOMAIN, config))
+
+    @patch('watchdog.observers')
```
@OttoWinter @dgomes I want to patch `watchdog.observers` but here I get the error `ModuleNotFoundError: No module named 'watchdog.observers'; 'watchdog' is not a package` which I don't understand as I have already mocked the watchdog package..?

robmarkcole committed:
Update test_file_watcher.py

robmarkcole committed:
Update test_file_watcher.py

robmarkcole committed:
Update test_file_watcher.py

robmarkcole committed:
Add event

robmarkcole commented on tests/components/test_file_watcher.py:
```diff
@@ -0,0 +1,88 @@
+        event_handler.process(fake_event)
+
+        self.hass.block_till_done()
+        last_event = self.events[-1]
```
@OttoWinter @dgomes any ideas why this fails?

balloob replied:
Why wouldn't you just pass a Mock instead of `hass` into `create_event_handler`. Then you can just read to be fired events from the mock?

balloob commented on .coveragerc:
```diff
@@ -94,6 +94,8 @@ omit =
     homeassistant/components/envisalink.py
     homeassistant/components/*/envisalink.py
 
+    homeassistant/components/folder_watcher.py
```
Since it doesn't have any platforms, just add it to the big list below.

robmarkcole committed:
Update test_file_watcher.py

robmarkcole commented on tests/components/test_file_watcher.py:
```diff
@@ -0,0 +1,89 @@
+                            'path': fake_event.src_path,
+                            'file': 'file.txt',
+                            'folder': 'test'}
+        fake_hass.bus.fire.assert_called_with(
```
This is not called, I don't know why?

balloob replied:
You check `if not event.is_directory`, so…

robmarkcole replied:
`fake_event.is_directory` is `False` so `bus.fire` should be called and does in my simplified external testing. 

I wonder if the issue is caused by `class EventHandler(PatternMatchingEventHandler)` since `PatternMatchingEventHandler` is a `MagicMock()`?

robmarkcole committed:
Update .coveragerc

robmarkcole committed:
Update test_file_watcher.py

robmarkcole commented on tests/components/test_file_watcher.py:
```diff
@@ -0,0 +1,95 @@
+                            'file': FILE,
+                            'folder': FOLDER}
+
+        fake_hass.bus.fire.assert_called()
```
Gives error: 
```
>       fake_hass.bus.fire.assert_called()

tests/components/test_folder_watcher.py:93:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

_mock_self = <Mock name='mock.bus.fire' id='4561351064'>

    def assert_called(_mock_self):
        """assert that the mock was called at least once
            """
        self = _mock_self
        if self.call_count == 0:
            msg = ("Expected '%s' to have been called." %
                   self._mock_name or 'mock')
>           raise AssertionError(msg)
E           AssertionError: Expected 'fire' to have been called.

/anaconda3/lib/python3.6/unittest/mock.py:786: AssertionError
```

robmarkcole commented:
@OttoWinter @balloob @dgomes have spent literally 2 hours trying to figure out why the final unit test fails, and I can't see why. The following is identical and passes OK:

```python
def create_event_handler(hass):
    """"Return the Watchdog EventHandler object."""

    class EventHandler():
        """Class for handling Watcher events."""

        def __init__(self, hass):
            """Initialise the EventHandler."""
            self.hass = hass

        def process(self, event):
            """On Watcher event, fire HA event."""
            if not event.is_directory:
                folder, file_name = os.path.split(event.src_path)
                self.hass.bus.fire(
                    DOMAIN, {
                        "event_type": event.event_type,
                        'path': event.src_path,
                        'file': file_name,
                        'folder': folder,
                        })

    return EventHandler(hass)
```
```python
fake_event = get_fake_event()
fake_hass = Mock()
event_handler = create_event_handler(fake_hass)
event_handler.process(fake_event)
fake_hass.bus.fire.assert_called()  # Passes
```

robmarkcole committed:
Update test_file_watcher.py

dgomes committed:
debug + join

dgomes committed:
test event

dgomes committed:
lint

dgomes committed:
lint

robmarkcole committed:
Merge pull request #3 from dgomes/folder_watch_tests

Folder watch tests

robmarkcole committed:
Rename test_file_watcher.py to test_folder_watcher.py

robmarkcole committed:
hound

robmarkcole commented on tests/components/test_folder_watcher.py:
```diff
@@ -0,0 +1,75 @@
+        """Check that HASS events are fired correctly on watchdog event."""
+        from watchdog.events import FileModifiedEvent
+
+        # cant use setup_component 
```
ok

robmarkcole committed:
Tidy test

robmarkcole committed:
Further refine test

robmarkcole committed:
Adds to test_all

robmarkcole committed:
Fix test for py35

robmarkcole commented:
In my tests if I have:
```python
self.assertTrue(self.hass.bus.fire.called)
```
python 3.5.3 test passes but python 3.6 fails, but if I have:
```python
self.hass.bus.fire.assert_called()
```
python 3.6 test passes but python 3.5.3 fails.
Please advise

robmarkcole committed:
Change test again

robmarkcole committed:
Update test_folder_watcher.py

robmarkcole committed:
Fix test

robmarkcole committed:
Merge remote-tracking branch 'origin/patch-5' into patch-5

# Conflicts:
#	tests/components/test_folder_watcher.py

robmarkcole committed:
Add watchdog to test

robmarkcole commented:
Travis is seriously driving me mad :-0

robmarkcole committed:
Update folder_watcher.py

robmarkcole committed:
add watchdog

robmarkcole committed:
Update folder_watcher.py

robmarkcole commented:
Request final review @OttoWinter @balloob

balloob commented:
🎉

bachulator commented:
Wondering is it easy enough to add network share (NFS, SAMBA, FTP etc) monitoring directly from config file. Would be super handy for prople using restricted environment like Hassio.

robmarkcole commented:
@bachulator I don’t think that’s covered by Watchdog

bachulator commented:
Ok thanks, sad though, from FS perspective is rather close functionality.
